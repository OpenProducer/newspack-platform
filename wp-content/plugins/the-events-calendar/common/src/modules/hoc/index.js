export { default as withStore } from './with-store';
export { default as withSaveData } from './with-save-data';
export { default as withForm } from './with-form';
export { default as withBlockCloser } from './with-block-closer';
export { default as withSelected } from './with-selected';
