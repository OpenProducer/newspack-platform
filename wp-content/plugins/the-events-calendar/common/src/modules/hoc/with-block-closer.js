/* eslint-disable max-len */
/**
 * External dependencies
 */
import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import { noop } from 'lodash';

const ESCAPE_KEY = 27;
export const EVENT_NAMESPACE = 'tribe:click:proxy';
export const dispatch = ( e ) => {
	e.target.dispatchEvent( new CustomEvent( EVENT_NAMESPACE, { bubbles: true } ) );
};
export const intercept = e => e.stopPropagation();

export default ( WrappedComponent ) => {
	/**
	 * Prevents clicks on block or blacklisted DOM elements
	 * from closing the block
	 *
	 * @class WithBlockCloser
	 * @extends {PureComponent}
	 */
	class WithBlockCloser extends PureComponent {
		static displayName = `WithBlockCloser( ${ WrappedComponent.displayName || WrappedComponent.name || 'Component ' }`

		static propTypes = {
			onClose: PropTypes.func.isRequired,
			classNameClickBlacklist: PropTypes.arrayOf( PropTypes.string ).isRequired,
			isOpen: PropTypes.bool.isRequired,
		};

		static defaultProps = {
			classNameClickBlacklist: [ '.edit-post-sidebar' ],
			onClose: noop,
			isOpen: false,
		}

		nodeRef = React.createRef();
		_eventNamespace = EVENT_NAMESPACE;

		/**
		 * dispatches custom events
		 *
		 * @memberof WithBlockCloser
		 * @param {Event} e event
		 */
		_dispatchClickProxyEvent = dispatch;

		// Prevent CustomEvents from propagating to document proxy listeners
		_interceptClickProxyEvent = intercept;

		/**
		 * keydown handler
		 *
		 * @memberof WithBlockCloser
		 * @param {Event} e event
		 */
		handleKeyDown = ( e ) => {
			if ( e.keyCode === ESCAPE_KEY ) {
				this.props.onClose();
			}
		}

		handleClick = () => this.props.onClose()

		componentDidMount() {
			this.props.isOpen && this._addEventListeners();
		}

		componentDidUpdate( prevProps ) {
			if ( prevProps.isOpen !== this.props.isOpen ) {
				this.props.isOpen
					? this._addEventListeners()
					: this._removeEventListeners();
			}
		}

		componentWillUnmount() {
			this._removeEventListeners();
		}

		get blacklistedNodes() {
			const classNames = this.props.classNameClickBlacklist.join( ', ' );
			return Array.from( document.querySelectorAll( classNames ) );
		}

		get node() {
			return this.nodeRef.current;
		}

		_addEventListeners() {
			// Intercept custom events bubbled in block or blacklisted nodes
			this.node.addEventListener( this._eventNamespace, this._interceptClickProxyEvent );
			this.blacklistedNodes.forEach(
				node => node.addEventListener( this._eventNamespace, this._interceptClickProxyEvent )
			);

			// Wait to receive custom events, if not intercepted, then go to click handler
			document.addEventListener( this._eventNamespace, this.handleClick );
			// Dispatch custom event on regular clicks
			document.addEventListener( 'click', this._dispatchClickProxyEvent );

			// Close on certain keypresses
			document.addEventListener( 'keydown', this.handleKeyDown );
		}

		_removeEventListeners() {
			this.node.removeEventListener( this._eventNamespace, this._interceptClickProxyEvent );
			this.blacklistedNodes.forEach(
				node => node.removeEventListener( this._eventNamespace, this._interceptClickProxyEvent )
			);

			document.removeEventListener( 'keydown', this.handleKeyDown );
			document.removeEventListener( this._eventNamespace, this.handleClick );
			document.removeEventListener( 'click', this._dispatchClickProxyEvent );
		}

		render() {
			return (
				<div ref={ this.nodeRef }>
					<WrappedComponent { ...this.props } />
				</div>
			);
		}
	}

	return WithBlockCloser;
};
