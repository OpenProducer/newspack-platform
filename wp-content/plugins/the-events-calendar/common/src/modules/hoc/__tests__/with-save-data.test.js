/**
 * External dependencies
 */
import renderer from 'react-test-renderer';
import React from 'react';

/**
 * Internal dependencies
 */
import { withSaveData } from '@moderntribe/common/hoc';

const Block = ( props ) => <div { ...props }>With Save Data!</div>;
const props = {
	name: 'tribe/event',
	setInitialState: jest.fn(),
	setAttributes: jest.fn(),
	title: 'Modern Tribe!',
	description: 'The Next Generation of Digital Agency',
	organizers: [],
	attributes: {
		title: 'Modern Tribe',
		description: '',
		organizers: [],
	},
};

describe( 'HOC - With Details', () => {
	let Wrapper;
	let component;
	let instance;

	beforeEach( () => {
		Wrapper = withSaveData()( Block );
		component = renderer.create( <Wrapper { ...props } /> );
		instance = component.root;
	} );

	afterEach( () => {
		props.setInitialState.mockClear();
		props.setAttributes.mockClear();
		component.getInstance().unregisterBlock();
	} );

	it( 'Should render a component', () => {
		expect( component.toJSON() ).toMatchSnapshot();
	} );

	it( 'Should render the inner component', () => {
		expect( instance ).not.toBe( null );
		expect( () => instance.findByType( Block ) ).not.toThrowError();
	} );

	it( 'Should set the initial state', () => {
		expect( props.setInitialState ).toHaveBeenCalled();
		expect( props.setInitialState ).toHaveBeenCalledTimes( 1 );
	} );

	it( 'Should generate the attributes', () => {
		const HOC = component.getInstance();
		expect( HOC.attrs ).toEqual( props.attributes );
	} );

	it( 'Should generate the keys', () => {
		const HOC = component.getInstance();
		expect( HOC.keys ).toEqual( Object.keys( props.attributes ) );
	} );

	it( 'Simulate componentDidUpdate call', () => {
		const wrapper = shallow( <Wrapper { ...props } /> );
		const wrapperInstance = wrapper.instance();
		expect( wrapperInstance.calculateDiff() ).toEqual( {
			description: 'The Next Generation of Digital Agency',
			title: 'Modern Tribe!',
		} );
		wrapper.setProps( {
			attributes: {
				title: 'Modern Tribe!',
				description: 'The Next Generation of Digital Agency',
				organizers: [],
			}
		} );
		expect( wrapperInstance.calculateDiff() ).toEqual( {} );
		wrapper.setProps( {
			organizers: [ 3 ],
			attributes: {
				title: 'Modern Tribe!',
				description: 'The Next Generation of Digital Agency',
				organizers: [ 3 ],
			},
		} );
		expect( wrapperInstance.calculateDiff() ).toEqual( {} );
		wrapper.setProps( {
			organizers: [ 2, 3 ],
			attributes: {
				title: 'Modern Tribe!',
				description: 'The Next Generation of Digital Agency',
				organizers: [ 3, 2 ],
			},
		} );
		expect( wrapperInstance.calculateDiff() ).toEqual( { organizers: [ 2, 3 ] } );
		wrapperInstance.unregisterBlock();
	} );

	it( 'Should calculate the diff', () => {
		const HOC = component.getInstance();
		const expected = {
			title: props.title,
			description: props.description,
		};
		expect( HOC.calculateDiff() ).toEqual( expected );
	} );

	it( 'Should count a single block', () => {
		const HOC = component.getInstance();
		expect( HOC.blockCount() ).toBe( 1 );
	} );
} );

describe( 'HOC - With Details on multiple instances', () => {
	afterEach( () => {
		props.setInitialState.mockClear();
		props.setAttributes.mockClear();
	} );

	it( 'Should register the initial state just once', () => {
		const WrapperComponent = withSaveData()( Block );

		renderer.create( <WrapperComponent { ...props } /> );
		renderer.create( <WrapperComponent { ...props } /> );
		renderer.create( <WrapperComponent { ...props } /> );

		expect( props.setInitialState ).toHaveBeenCalled();
		expect( props.setInitialState ).toHaveBeenCalledTimes( 1 );
	} );

	it( 'Should register the state multiple times on non isolated instances', () => {
		const WrapperComponent = withSaveData()( Block );
		props.isolated = true;
		renderer.create( <WrapperComponent { ...props } /> );
		renderer.create( <WrapperComponent { ...props } /> );
		renderer.create( <WrapperComponent { ...props } /> );

		expect( props.setInitialState ).toHaveBeenCalled();
		expect( props.setInitialState ).toHaveBeenCalledTimes( 3 );
	} );
} );

describe( 'HOC - test life cycle callbacks', () => {
	let Wrapper;
	let component;
	let instance;
	let properties = {};

	beforeAll( () => {
		props.onBlockCreated = jest.fn();
		props.onBlockRemoved = jest.fn();
	});

	beforeEach( () => {
		Wrapper = withSaveData()( Block );
		component = renderer.create( <Wrapper { ...props } /> );
		instance = component.root;
		properties = instance.props;
	} );

	afterEach( () => {
		props.setInitialState.mockClear();
		props.setAttributes.mockClear();
		props.onBlockCreated.mockClear();
		props.onBlockRemoved.mockClear();
	} );

	afterAll( () => {
		delete props.onBlockCreated;
		delete props.onBlockRemoved;
	} );

	it( 'Should call the onBlockCreated callback on mount', () => {
		expect( props.onBlockCreated ).toHaveBeenCalled();
		expect( props.onBlockCreated ).toHaveBeenCalledWith( properties );
		expect( props.onBlockRemoved ).not.toHaveBeenCalled();
	} );

	it( 'Should call the onBlockRemoved callback on unmount of the block', () => {
		component.unmount();
		expect( props.onBlockCreated ).toHaveBeenCalled();
		expect( props.onBlockCreated ).toHaveBeenCalledWith( properties );
		expect( props.onBlockRemoved ).toHaveBeenCalled();
		expect( props.onBlockRemoved ).toHaveBeenCalledWith( properties );
	} );
} );
