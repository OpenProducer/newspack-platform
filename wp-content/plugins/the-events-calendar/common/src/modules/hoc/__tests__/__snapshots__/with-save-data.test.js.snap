// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HOC - With Details Should render a component 1`] = `
<div
  attributes={
    Object {
      "description": "",
      "organizers": Array [],
      "title": "Modern Tribe",
    }
  }
  description="The Next Generation of Digital Agency"
  isolated={false}
  name="tribe/event"
  onBlockCreated={[Function]}
  onBlockRemoved={[Function]}
  organizers={Array []}
  setAttributes={[MockFunction]}
  setInitialState={
    [MockFunction] {
      "calls": Array [
        Array [
          Object {
            "attributes": Object {
              "description": "",
              "organizers": Array [],
              "title": "Modern Tribe",
            },
            "description": "The Next Generation of Digital Agency",
            "get": [Function],
            "isolated": false,
            "name": "tribe/event",
            "onBlockCreated": [Function],
            "onBlockRemoved": [Function],
            "organizers": Array [],
            "setAttributes": [MockFunction],
            "setInitialState": [MockFunction] {
              "calls": [Circular],
              "results": Array [
                Object {
                  "isThrow": false,
                  "value": undefined,
                },
              ],
            },
            "title": "Modern Tribe!",
          },
        ],
      ],
      "results": Array [
        Object {
          "isThrow": false,
          "value": undefined,
        },
      ],
    }
  }
  title="Modern Tribe!"
>
  With Save Data!
</div>
`;
